File: web_frontend/frontend/src/api/agentHub.ts
import { requestJson, withQuery } from './http'

export interface AgentHubUser {
  user_id: string
  github_login: string
  display_name: string
  avatar_url: string
  is_admin: boolean
}

export interface AgentHubAuthStatus {
  authenticated: boolean
  user: AgentHubUser | null
  hub_url: string
}

export interface AgentHubBrowserAuthorization {
  flow_id: string
  poll_secret: string
  authorization_url: string
  expires_in: number
  interval: number
}

export interface AgentHubValidation {
  dependencies?: {
    python_count?: number
    npm_count?: number
    system_count?: number
    python?: string[]
    npm?: string[]
    system?: string[]
  }
  tools?: {
    package_tools?: string[]
    mcp_servers?: string[]
    builtin_tools?: string[]
  }
  warnings?: Array<{ code: string; message: string; path?: string }>
}

export interface AgentHubRelease {
  release_id: string
  publisher: string
  package_id: string
  name: string
  description: string
  version: string
  sha256: string
  size_bytes: number
  status: string
  validation: AgentHubValidation | null
  download_count: number
  published_at: string
}

export interface AgentHubPackageList {
  items: AgentHubRelease[]
  total: number
  limit: number
  offset: number
}

export interface AgentHubUpload {
  upload_id: string
  filename: string
  expected_size: number
  actual_size: number | null
  status: string
  error: { code: string; message: string } | null
  validation: AgentHubValidation | null
  created_at: string
  updated_at: string
}

export interface AgentHubSkillFileDraft {
  path: string
  size_bytes: number
  kind: 'text' | 'binary'
  included: boolean
  content: string | null
}

export interface AgentHubSkillDraft {
  skill_id: string
  source: string
  enabled: boolean
  required: boolean
  path: string
  files: AgentHubSkillFileDraft[]
}

export interface AgentHubPublishPreview {
  package_id: string
  mcp_servers: Record<string, unknown>
  skills: AgentHubSkillDraft[]
}

export const agentHubApi = {
  auth: () => requestJson<AgentHubAuthStatus>('/api/agent-hub/auth'),
  startBrowserLogin: () =>
    requestJson<AgentHubBrowserAuthorization>('/api/agent-hub/auth/browser/start', { method: 'POST' }),
  pollBrowserLogin: (flowId: string, pollSecret: string) =>
    requestJson<{ status: string; retry_after_seconds?: number; user?: AgentHubUser }>(
      '/api/agent-hub/auth/browser/poll',
      {
        method: 'POST',
        body: JSON.stringify({ flow_id: flowId, poll_secret: pollSecret }),
      },
    ),
  cancelBrowserLogin: (flowId: string, pollSecret: string) =>
    requestJson<{ status: 'cancelled' }>('/api/agent-hub/auth/browser/cancel', {
      method: 'POST',
      body: JSON.stringify({ flow_id: flowId, poll_secret: pollSecret }),
    }),
  logout: () => requestJson<{ authenticated: false }>('/api/agent-hub/auth/logout', { method: 'POST' }),
  packages: (query = '', limit = 40, offset = 0) =>
    requestJson<AgentHubPackageList>(
      withQuery('/api/agent-hub/packages', { q: query, limit, offset }),
    ),
  uploads: (limit = 50) =>
    requestJson<AgentHubUpload[]>(withQuery('/api/agent-hub/uploads', { limit })),
  install: (releaseId: string, replace = false) =>
    requestJson<{ release: AgentHubRelease; package: Record<string, unknown> }>(
      `/api/agent-hub/releases/${encodeURIComponent(releaseId)}/install`,
      {
        method: 'POST',
        body: JSON.stringify({ replace }),
      },
    ),
  publishPreview: (packageId: string) =>
    requestJson<AgentHubPublishPreview>(
      `/api/agent-hub/packages/${encodeURIComponent(packageId)}/publish-preview`,
    ),
  publish: (
    packageId: string,
    extensions: {
      mcp_servers: Record<string, unknown>
      skills: Array<{
        skill_id: string
        files: Array<{ path: string; included: boolean; content: string | null }>
      }>
    },
  ) =>
    requestJson<AgentHubUpload>(
      `/api/agent-hub/packages/${encodeURIComponent(packageId)}/publish`,
      {
        method: 'POST',
        body: JSON.stringify({
          confirmed_sensitive_review: true,
          extensions,
        }),
      },
    ),
}

File: web_frontend/backend/routes/agent_hub.py
from __future__ import annotations

import asyncio
from collections.abc import Callable
from typing import Any

from fastapi import APIRouter, HTTPException, Query

from agent_factory.factory_graph.frontend_bridge.agent_package_runtime import (
    AgentPackageRuntimeManager,
)
from web_frontend.backend.agent_hub_client import AgentHubClient, AgentHubClientError


def create_agent_hub_router(
    *,
    client: AgentHubClient | None = None,
    runtime_factory: Callable[[], AgentPackageRuntimeManager] | None = None,
) -> APIRouter:
    router = APIRouter(prefix="/api/agent-hub")
    hub = client or AgentHubClient()
    resolve_runtime = runtime_factory or AgentPackageRuntimeManager

    @router.get("/auth")
    async def auth_status():
        return await _call(hub.auth_status)

    @router.post("/auth/browser/start")
    async def start_browser_login():
        return await _call(hub.start_browser_login)

    @router.post("/auth/browser/poll")
    async def poll_browser_login(payload: dict[str, Any]):
        flow_id = str(payload.get("flow_id") or "").strip()
        poll_secret = str(payload.get("poll_secret") or "").strip()
        if not flow_id or not poll_secret:
            raise HTTPException(status_code=422, detail="browser login flow is required")
        return await _call(
            hub.poll_browser_login,
            flow_id=flow_id,
            poll_secret=poll_secret,
        )

    @router.post("/auth/browser/cancel")
    async def cancel_browser_login(payload: dict[str, Any]):
        flow_id = str(payload.get("flow_id") or "").strip()
        poll_secret = str(payload.get("poll_secret") or "").strip()
        if flow_id and poll_secret:
            await _call(
                hub.cancel_browser_login,
                flow_id=flow_id,
                poll_secret=poll_secret,
            )
        return {"status": "cancelled"}

    @router.post("/auth/logout")
    async def logout():
        return await _call(hub.logout)

    @router.get("/packages")
    async def list_packages(
        q: str = Query(default="", max_length=200),
        limit: int = Query(default=20, ge=1, le=100),
        offset: int = Query(default=0, ge=0),
    ):
        return await _call(hub.list_packages, query=q, limit=limit, offset=offset)

    @router.get("/packages/{publisher}/{package_id}")
    async def package_detail(publisher: str, package_id: str):
        return await _call(hub.package_detail, publisher, package_id)

    @router.get("/uploads")
    async def list_uploads(limit: int = Query(default=50, ge=1, le=100)):
        return await _call(hub.list_uploads, limit=limit)

    @router.post("/packages/{package_id}/publish")
    async def publish_package(package_id: str):
        return await _call(
            hub.publish_package,
            package_id,
            runtime=resolve_runtime(),
        )

    @router.post("/releases/{release_id}/install")
    async def install_release(release_id: str, payload: dict[str, Any] | None = None):
        return await _call(
            hub.install_release,
            release_id,
            replace=bool((payload or {}).get("replace")),
            runtime=resolve_runtime(),
        )

    return router


async def _call(function, *args, **kwargs):
    try:
        return await asyncio.to_thread(function, *args, **kwargs)
    except AgentHubClientError as exc:
        raise HTTPException(
            status_code=exc.status_code,
            detail={"code": exc.code, "message": str(exc)},
        ) from exc
    except FileExistsError as exc:
        raise HTTPException(
            status_code=409,
            detail={"code": "package_already_installed", "message": str(exc)},
        ) from exc
    except FileNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc
